### ISSUE TITLE
.Net Core implementation that works with Lambda Serverless?

### ISSUE THREAD

**Issue by bmatthews99** [AreYouFreeBusy/AlexaSkillsKit.NET#20]
One of the big selling points for the ease of Alexa dev is how easy (and cheap) it integrates with lambda. Also some orgs are moving to Core as it becomes more stable. Any future plans to have this work with .Net Core for those easy Lambda projects?

**Comment by MaxHorstmann** [comment]
+1 would be great to get this to work with the recently added C# support in Lambda

**Comment by goldfishrock** [comment]
Another +1.  I've been using the Slight.Alexa version thats been ported to Core... but it's incomplete at the moment.

**Comment by tonysurma** [comment]
I have a .Net core (netstandard) port of this [in a fork](https://github.com/tonysurma/AlexaSkillsKit.NET/tree/dotnetcore) that initially is working for me but I will test further before declaring success beyond *It works on my machine* ™️  - I will also verify it works in core on Azure and AWS Lambda

**Comment by tonysurma** [comment]
verified it is working on azure just now, up next AWS Lambda verification (but if it works on .net core it should on another)

**Comment by szymczakk** [comment]
@tonysurma I couldn't find any nuget feed with your version of this lib (for .net core) so I allowed myself to create one: https://www.myget.org/feed/ami/package/nuget/AlexaSkillsKit.Net (if any of you need it, you can reference it). Quite useful, because local dll can't be referenced in easy way in VSCode.

**Comment by stefann42** [comment]
@szymczakk Happy to amend the "official" NUGet package definition so it can be used from .NET Core if you can educate me a bit about what I need to do. (My head is still on .NET Framework 4.6 and I haven't had time to learn about porting stuff to .NET Core.) Is the package definition you published on MyGet available somewhere for me to look at?

**Comment by tonysurma** [comment]
@stefann42 yes it is!  I moved to 1.6 (no reason other than it needed to be a different version.  in .NET Core it comes from the csproj now so you can see the details [here ](https://github.com/tonysurma/AlexaSkillsKit.NET/blob/dotnetcore/AlexaSkillsKit.Lib/AlexaSkillsKit.Lib.csproj)(I literally copied over what you had done from the nuspec to the csproj).

I also chose the right netstandard so it will work with .NET Core and the version of .NET Framework you had previously targeted so it is compatible and should work without change in existing .NET 4.6 deployments.

The changes are all in the [fork ](https://github.com/tonysurma/AlexaSkillsKit.NET/tree/dotnetcore) I made and I strove to make each commit 'small' so it documented the porting.  Happy to help however I can and if you would like it I can make a PR with these changes and anything else you would need to convert.

**Comment by tonysurma** [comment]
thank you @szymczakk for posting that.  I hadn't gotten to that step yet.  I had been using a custom nuget feed entry that pointed to a local folder where I had the package (simple but it worked for me)

**Comment by szymczakk** [comment]
@tonysurma  How did you handle the request? When I create API project, and in controller get the Request it's type ```Microsoft.AspNetCore.Http.HttpRequest``` and **Speechlet** in method ```GetResponse``` require ```System.Net.Http.HttpRequestMessage``` and I don't know how to convert it. I was thinking about rewriting ```GetRespose``` method to use request from .net Core, what do you think?

**Comment by tonysurma** [comment]
@szymczakk I made a `SampleCore ` project [here](https://github.com/tonysurma/AlexaSkillsKit.NET/tree/dotnetcore/AlexaSkillsKit.SampleCore) to demonstrate that and included commit(s) that showed each step to 'upgrade' the Sample to Core.  The trick for that I recall is [including](https://github.com/tonysurma/AlexaSkillsKit.NET/commit/da3dc204d928093cd5b25c7ae0b197dd26109d63) `WebApiCompatShim` which allows you to inherit from `ApiController` and aligns to the types expected by the Speechlet API

I did that as it was the quickest/most direct port.  I was also thinking of adding another `GetResponse` method/overload to adapt to the .net core approach but recall there were some items that the Speechlet wanted that would be a bit messy to get.

**Comment by szymczakk** [comment]
@tonysurma sory for disturbing you, but have another question, can you [look here](https://github.com/szymczakk/AMI.AlexaAPI/tree/master/AMI.AlexaAPI) and see what do I do wrong? By wrong I mean 

> The remote endpoint could not be called, or the response it returned was invalid.

as an response from service simulator in Skill config console on AWS. I got request, but response that I send to Alexa is *invalid* 😢

**Comment by tonysurma** [comment]
@szymczakk no worries, I will try to look at that a bit later.  Can you see the response you are sending from your code and are you sure you got called? I ask only as I have gotten that error when tunneling or some other url issue was the cause.

**Comment by szymczakk** [comment]
@tonysurma Yep, Im 100% sure that my server is called, because I'm attached with debugger to it, and see everything is happening. I will try to catch my whole response with Fiddler and show you ;)

**Comment by tonysurma** [comment]
got it @szymczakk - just looked through the code and nothing stood out (looked at startup, controller, speechlet) and all looks good with no gotchas.
I also reverified my deployment with this version is working as well as made a quick simple one and it worked for me.  Curious what the response is that it doesn't like.  Sorry I couldn't be more helpful (yet)

**Comment by szymczakk** [comment]
@tonysurma Thats weird. I'm unable to capture my trafic with Fiddler, but I can stop execution by debugger (so my local server is called). How did you configure your server? Do you use IIS? How about certificate? Slefsigned or trusted? And do you have static IP address or you use some DynDNS (like no-ip etc) ?

**Comment by tonysurma** [comment]
It's been a bit since I did the self hosting (I do updates now to deployed instances in Azure App Service websites) but I recall following this guide.  I recall getting tripped up a couple of time and used this [guide ](https://freebusy.io/blog/getting-started-with-alexa-app-development-for-amazon-echo-using-dot-net) and another I can't find right now.  Eventually I went to just using the site on azure and logging and other mechanisms as it was faster (especially since I tend to dev on combination of desk pc and a couple of laptops).  Sorry I don't have any more insights there

**Comment by szymczakk** [comment]
Ok, I found the response: 

> {"endpointRegion":"EU","text":"hello","applicationId":"<appId>","locale":"en_US","session":null}

**Comment by tonysurma** [comment]
hmmm, let me look and compare - hopefully there is something there

**Comment by szymczakk** [comment]
@tonysurma I compared it with [amazon specification](https://developer.amazon.com/public/solutions/alexa/alexa-skills-kit/docs/alexa-skills-kit-interface-reference) and it's problem with respone :/ Ehhh... have to figure it out somehow 😉

**Comment by tonysurma** [comment]
gotcha - yeah.

My response from an app on the same build of the framework looks like this (removing a few specific bits)

`{
  "version": "1.0",
  "response": {
    "outputSpeech": {
      "type": "PlainText",
      "text": "My content"
    },
    "shouldEndSession": true
  },
  "sessionAttributes": {
    "intentSequence": "MyIntentId"
  }
}`

Definitely looks like stuff is missing in the response building.

**Comment by szymczakk** [comment]
Hey @tonysurma ! Sory for taking it so long, but have some more info, maybe you will know how to handle it. When I debug the request it shows that the response that is generated by code is 

> {"response":{"card":{"type":"Simple","content":"SessionSpeechlet - Hello","title":"SessionSpeechlet - Title","subtitle":"SessionSpeechlet - Sub Title"},"outputSpeech":{"type":"PlainText","text":"Hello"},"shouldEndSession":true},"sessionAttributes":{"intentSequence":"HelloWorldIntent"},"version":"1.0"}


but when I finally cought the correct resposne that is sent by server it turns out that it's 

> {
  "version": {
    "major": 1,
    "minor": 1,
    "build": -1,
    "revision": -1,
    "majorRevision": -1,
    "minorRevision": -1
  },
  "content": {
    "headers": [
      {
        "key": "Content-Type",
        "value": [
          "application/json; charset=utf-8"
        ]
      }
    ]
  },
  "statusCode": 200,
  "reasonPhrase": "OK",
  "headers": [],
  "requestMessage": null,
  "isSuccessStatusCode": true
}

I run dotnet core in kestrel and then host it on IIS on windows 10. Do you have any idea why do I got such resposne?

**Comment by tonysurma** [comment]
Sorry but unfortunately I do not. It looks to be like it's encoding the http response itself in json which is strange and i've never seen before myself.

**Comment by szymczakk** [comment]
Hey @tonysurma, look [what I found](https://stackoverflow.com/questions/37429190/asp-net-core-httprequestmessage-returns-strange-json-message). And this link is explaining why it wasn't working. But funny case is that your code works for you, but you don't have [this](https://github.com/szymczakk/AMI.AlexaAPI/commit/f4d8b7a3d236849fd3776e40dfb67086fa70a7b8#diff-ee0b4f76d4d60bd5ebb81af72339d1b6R31) function in your source.

**Comment by tonysurma** [comment]
Interesting, great find.  I will take a look. Might be an omission from the sample project I had in my test and production projects... thanks

**Comment by tonysurma** [comment]
Thanks again @szymczakk I had that edit in each of my projects and missed it in the SampleCore project as I never 'directly' ran the sample.  I was using the SampleCore project as the basis for my projects to test and run and forgot to 'back port' that line into the sample.  I pushed this [now](https://github.com/tonysurma/AlexaSkillsKit.NET/commit/f2edad311765bd6d2572188bccda1e7b1b089110).

**Comment by JustMaier** [comment]
Hey @tonysurma and @stefann42 any chance we could still get this merged and published to nuget? Now that we've got .net core 2.0, it'd be great to get .net core support.

**Comment by tonysurma** [comment]
@JustMaier if @stefann42 is open to it, I can update/test my fork with netstandard 2 to ready it for merge

**Comment by stefann42** [comment]
@tonysurma By all means. I'm happy to accept the PR, just let me know if I need to be aware of any specifics. We don't yet use .NET Core in our production deployment so I don't have an environment to test myself.

**Comment by ElvenMonky** [comment]
Now, when interfaces support is merged, I believe .Net Core support is our main priority. @tonysurma changes in your fork looks great to me. Could you, please, create a PR with your changes, so that we can include .Net Core support to the nearest library release? Just tell if you need any help in preparing PR or merging with latest changes.

**Comment by ElvenMonky** [comment]
One thing to note. Current library is compatible with VS2012 for both Speechlet authors and AlexaSkillsKit.Net contributors. [.Net Standard](https://docs.microsoft.com/en-us/dotnet/standard/net-standard) version will require .Net Core1.x SDK and VS 2017 or VS Code for development. @stefann42 are you ready for these kind of changes in our development environment?

For Speechlet authors we can provide multitargeting NuGet package, ofcourse.

**Comment by stefann42** [comment]
I think we should start a separate branch for version 2 and make that compatible with .NET Core. We’re not ready to move off .NET Framework 4.7.

**Comment by ElvenMonky** [comment]
I have slightly different proposal, which I've tried to describe in newly created wiki [Roadmap](https://github.com/AreYouFreeBusy/AlexaSkillsKit.NET/wiki/Roadmap#net-core-support) document.

**Comment by stefann42** [comment]
@ElvenMonky Love the Roadmap! Let's have an online meeting together next week (between Christmas and New Year's) and discuss merged PRs and strategy for moving forward. Please see https://freebusy.link/stefan for my availability and to schedule the meeting.

**Comment by tonysurma** [comment]
thanks @ElvenMonky for noting two issues with my fork.  I just addressed those and will also bring it up to date with this repository recent commits.  Happy to help as you all decide on how to address .net core 2

**Comment by ElvenMonky** [comment]
Well. currently I think we can drop VS 2012 support, after already implemented code is released.
@stefann42 Probably I'll reschedule for your 27th evening (my 28th morning).

**Comment by ElvenMonky** [comment]
@tonysurma we created separate branch for .Net Framework version. Please, feel free to create PR into the master branch with your changes related to .Net Standard/.Net Core support.

**Comment by brandonh-msft** [comment]
the myget URL posted above doesn't work. Is it still alive?
~~~
[ask.net myget] 'doctype' is an unexpected token. The expected token is 'DOCTYPE'. Line 1, position 3.
~~~
in VS 2017.15.6.6

**Comment by szymczakk** [comment]
@brandonh-msft Do you mean this one https://www.myget.org/feed/ami/package/nuget/AlexaSkillsKit.Net/1.7.0 ?? This one is still working (just checked it)

**Comment by Zoba273** [comment]
is this branch still alive?

**Comment by ElvenMonky** [comment]
Development in this repository is inactive since January 2018. It seems nobody is interested.
This is a 3rd party library, while Java version is supported officially by Amazon.
